File: book/petrol-pump/java/chapter4/section6/AccumulatePulsesPump.java
package chapter4.section6;

import pump.*;
import chapter4.section4.LifeCycle;
import nz.sodium.*;
import java.util.Optional;

public class AccumulatePulsesPump implements Pump {
    public Outputs create(Inputs inputs) {
        LifeCycle lc = new LifeCycle(inputs.sNozzle1,
                                     inputs.sNozzle2,
                                     inputs.sNozzle3);
        Cell<Double> litersDelivered =
                accumulate(lc.sStart.map(u -> Unit.UNIT),
                           inputs.sFuelPulses,
                           inputs.calibration);
        return new Outputs()
            .setDelivery(lc.fillActive.map(
                of ->
                    of.equals(Optional.of(Fuel.ONE))   ? Delivery.FAST1 :
                    of.equals(Optional.of(Fuel.TWO))   ? Delivery.FAST2 :
                    of.equals(Optional.of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF))
            .setSaleQuantityLCD(litersDelivered.map(
                    q -> Formatters.formatSaleQuantity(q)));
    }

    public static Cell<Double> accumulate(
            Stream<Unit> sClearAccumulator,
            Stream<Integer> sPulses,
            Cell<Double> calibration) {
        CellLoop<Integer> total = new CellLoop<>();
        total.loop(sClearAccumulator.map(u -> 0)
               .orElse(
                   sPulses.snapshot(total, (pulses_, total_) ->
                       pulses_ + total_)
               )
               .hold(0));
        return total.lift(calibration,
            (total_, calibration_) -> total_ * calibration_);
    }
}

File: book/petrol-pump/c#/PetrolPump/Chapter4/Section6/AccumulatePulsesPump.cs
using PetrolPump.Chapter4.Section4;
using Sodium.Frp;
using Sodium.Functional;

namespace PetrolPump.Chapter4.Section6
{
    public class AccumulatePulsesPump : IPump
    {
        public Outputs Create(Inputs inputs)
        {
            LifeCycle lc = new LifeCycle(inputs.SNozzle1,
                inputs.SNozzle2,
                inputs.SNozzle3);
            Cell<double> litersDelivered =
                Accumulate(lc.SStart.Map(_ => Unit.Value),
                    inputs.SFuelPulses,
                    inputs.Calibration);
            return new Outputs()
                .SetDelivery(lc.FillActive.Map(
                    m =>
                        m.Equals(Maybe.Some(Fuel.One)) ? Delivery.Fast1 :
                            m.Equals(Maybe.Some(Fuel.Two)) ? Delivery.Fast2 :
                                m.Equals(Maybe.Some(Fuel.Three)) ? Delivery.Fast3 :
                                    Delivery.Off))
                .SetSaleQuantityLcd(litersDelivered.Map(Formatters.FormatSaleQuantity));
        }

        public static Cell<double> Accumulate(
            Stream<Unit> sClearAccumulator,
            Stream<int> sPulses,
            Cell<double> calibration)
        {
            CellLoop<int> total = new CellLoop<int>();
            total.Loop(sClearAccumulator.Map(u => 0)
                .OrElse(sPulses.Snapshot(total, (pulsesLocal, totalLocal) => pulsesLocal + totalLocal))
                .Hold(0));
            return total.Lift(
                calibration,
                (totalLocal, calibrationLocal) => totalLocal * calibrationLocal);
        }
    }
}